src/climax/enso/datamodule.py

```diff
@@ -0,0 +1,99 @@
+from datetime import date, timedelta
+
+from torch.utils.data import DataLoader
+from pytorch_lightning import LightningDataModule
+import xarray as xr
+
+from climax.enso.dataset import ENSODataset
+
+
+class ENSODataModule(LightningDataModule):
+    def __init__(
+        self,
+        data_dir: str,
+        start_train_year: int,
+        start_val_year: int,
+        start_test_year: int,
+        end_year: int,
+        forecast_lead_time: int,
+        batch_size: int,
+        num_workers: int = 0,
+        pin_memory: bool = False,
+    ):
+        super().__init__()
+
+        # Dataset args
+        self.data_dir = data_dir
+        self.start_train_year = start_train_year
+        self.start_val_year = start_val_year
+        self.start_test_year = start_test_year
+        self.end_year = end_year
+        self.forecast_lead_time = forecast_lead_time
+
+        # Dataloader args
+        self.batch_size = batch_size
+        self.num_workers = num_workers
+        self.pin_memory = pin_memory
+
+        # Set up datasets
+        # TODO: is here the best place to do this?
+        # TODO: fix paths
+        era5_dataset = xr.load_dataset(self.data_dir + "/era5_daily.nc")
+        enso_dataset = xr.load_dataset(self.data_dir + "/enso_daily.nc")
+        self.train_dataset = self.get_dataset(
+            era5_dataset,
+            enso_dataset,
+            self.start_train_year,
+            self.start_val_year - 1,
+        )
+        self.val_dataset = self.get_dataset(
+            era5_dataset,
+            enso_dataset,
+            self.start_val_year,
+            self.start_test_year - 1,
+        )
+        self.test_dataset = self.get_dataset(
+            era5_dataset,
+            enso_dataset,
+            self.start_test_year,
+            self.end_year,
+        )
+
+    def setup(self, stage):
+        pass
+
+    def get_dataset(self, era5_dataset, enso_dataset, start_year, end_year):
+        start_date = date(start_year, 1, 1)
+        end_date = date(end_year, 12, 31)
```

Comment:
You removed the leap year treatment, right? I also think we can deal with leap years. Applying the transformers blocks to the time series with one element more should also work. What do you think about the time embedding in this case? The climate projection model has a learned time embedding (the lead time embedding is separate). I am not quite sure how to deal with the leap year there.

Reply:
Since we are dealing with daily data, I don't actually understand why we need to give the leap days special treatment?

Reply:
Not really special treatment, but it's just a bit unclear how best to do the embedding with it. The climate projection architecture that we could use has a learned embedding for each time step. https://github.com/maximilian-gelbrecht/ClimaX/blob/c039875258a493156c38c878cc62472139f47ed2/src/climax/climate_projection/arch.py#L57

With some inputs being one element longer than others, this just becomes a bit unclear how to do it best. But I think we can also think about some way how to deal with it.

Reply:
Ah yes, sorry, I was somehow thinking of always giving one year (somehow including leap years). But you are totally right, just X days is absolutely the way to go.

Reply:
Makes we wonder why they even treated leap years differently in the paper in the first place

Reply:
Thinking a bit more about this, in ENSO prediction, there are certain times in the year when it's harder to forecast then in others. So the seasonal dynamics play a role. If I recall correctly, e.g. there's a "spring barrier" through which predictability goes down. So we could think about the temporal embedding still encoding the calendar day in some way. Maybe. But yeah, I thinks that's just one of the many things to play around with.